src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,566 @@
+ * constraints:
+ * <ul>
+ *     <li>Gatherers whose initializer is {@link #defaultInitializer()} are
+ *     considered to be stateless, and invoking their initializer is optional.
+ *     </li>
+ *     <li>Gatherers whose integrator is an instance of {@link Integrator.Greedy}
+ *     can be assumed not to short-circuit, and the return value of invoking
+ *     {@link Integrator#integrate(Object, Object, Downstream)} does not need to
+ *     be inspected.</li>
+ *     <li>The first argument passed to the integration function, both
+ *     arguments passed to the combiner function, and the argument passed to the
+ *     finisher function must be the result of a previous invocation of the
+ *     initializer, integrator, or combiner functions.</li>
+ *     <li>The implementation should not do anything with the result of any of
+ *     the initializer, integrator, or combiner functions other than to
+ *     pass them again to the integrator, combiner, or finisher functions.</li>
+ *     <li>Once a state object is passed to the combiner or finisher function,
+ *     it is never passed to the integrator function again.</li>
+ *     <li>When the integrator function returns {@code false},
+ *     it shall be interpreted just as if there were no more elements to pass
+ *     it.</li>
+ *     <li>For parallel evaluation, the gathering implementation must manage
+ *     that the input is properly partitioned, that partitions are processed
+ *     in isolation, and combining happens only after integration is complete
+ *     for both partitions.</li>
+ *     <li>Gatherers whose combiner is {@link #defaultCombiner()} may only be
+ *     evaluated sequentially. All other combiners allow the operation to be
+ *     parallelized by initializing each partition in separation, invoking
+ *     the integrator until it returns {@code false}, and then joining each
+ *     partitions state using the combiner, and then invoking the finisher on
+ *     the joined state. Outputs and state later in the input sequence will
+ *     be discarded if processing an earlier segment short-circuits.</li>
+ *     <li>Gatherers whose finisher is {@link #defaultFinisher()} are considered
+ *     to not have an end-of-stream hook and invoking their finisher is
+ *     optional.</li>
+ * </ul>
+ *
+ * @see Stream#gather(Gatherer)
+ * @see Gatherers
+ *
+ * @param <T> the type of input elements to the gatherer operation
+ * @param <A> the potentially mutable state type of the gatherer operation
+ *            (often hidden as an implementation detail)
+ * @param <R> the type of output elements from the gatherer operation
+ * @since 22
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+public interface Gatherer<T, A, R> {
+    /**
+     * A function that produces an instance of the intermediate state used for
+     * this gathering operation.
+     *
+     * <p>By default, this method returns {@link #defaultInitializer()}
```

Comment:
default methods usually have `@implSpec` for the text that specifies what the default implementation does.

Reply:
@AlanBateman Sounds good, I'll make it an implSpec.